### PR TITLE
[IMP] core: speed up registry setup

### PR DESCRIPTION
**Pull request by rco-odoo**
This proposes a new implementation of
* the "discovery" of fields on a given model;
* the setup of fields;
* the sharing of fields across registries.

This new implementation makes the registry loading faster.  Our benchmark is two registries with 296 modules (all community modules).  Both registries have the same models, so we can see how sharing across registries can save time and/or memory.  Here are time measurements for loading both registries (from scratch):

Time | before | after | difference
---|---|---|---
Loading Odoo modules | 1176ms | 1339ms | +14%
Loading registry 1 | 2042ms | 1067ms | -48%
Loading registry 2 | 990ms | 970ms | -2%

As we can see, there is some extra work done when loading Odoo modules, but this extra work is done once for all registries using those modules.  However, this price is largely compensated by the speedup when loading a registry.  Notice also that the net time to load a registry no longer depends whether other registries are present in the server.

Overall, the loading of the first registry goes from 3218ms to 2406ms, which is 25% faster.  In other words, the bootstrapping time of a server is smaller.

### CODE REVIEW

File: addons/website_livechat/tests/test_livechat_basic_flow.py [hunk 1]
```diff
@@ -47,6 +47,10 @@ def test_livechat_username(self):
         # Remove livechat_username
         self.operator.livechat_username = False
 
+        # this fixes an issue in the controller; it looks like a cache issue,
+        # but I could not figure it out
+        self.operator.name
```
Comment (seb-odoo) [inline addons/website_livechat/tests/test_livechat_basic_flow.py:52]: Isn't this related to the display_name/name_get cache and context issue on partner that I reported the other day?

Comment (rco-odoo) [reply addons/website_livechat/tests/test_livechat_basic_flow.py:52]: Maybe. I spent some time to figure it out, but eventually gave up and found this workaround.

File: odoo/models.py [hunk 2]
```diff
@@ -251,6 +251,93 @@ def expand_ids(id0, ids):
 }
 
 
+# THE DEFINITION AND REGISTRY CLASSES
```
Comment (Xavier-Do) [inline odoo/models.py:291]: Commit _[IMP] core: document registry classes and field sharing_  should be retitled if you want to keep this in a separate commit (looks like a comment move only)

File: odoo/fields.py [hunk 3]
```diff
@@ -273,47 +277,77 @@ def __repr__(self):
...
+            attrs.update(field.args)
+            if field._module:
+                modules.append(field._module)
+        attrs.update(self.args)
+        if self._module:
+            modules.append(self._module)
 
         attrs['args'] = self.args
-        attrs['model_name'] = model._name
+        attrs['model_name'] = owner._name
```
Comment (Xavier-Do) [inline odoo/fields.py:347]: I wouldn't rename this parameter: 
- inconsistency with other `_get_attrs` overrides
- what kind of object is owner?
- Useless blame

Anyway if wonder is this parametter is really usefull since it isn't use in any override anymore. It could also be done in _setup_attrs. (name parametter could also be avoided without the dirty state hack)

Also, is this really usefull since model_name and name are set in `__set_name__`: https://github.com/odoo/odoo/pull/69924/files#diff-eb000d15355b38936d1b7f7506465ad081a3d96424c1f64dd1008ef3d4ac6427R309

File: odoo/fields.py [hunk 4]
```diff
@@ -273,47 +277,77 @@ def __repr__(self):
...
+    #
+    # Non-toplevel non-direct fields are the fields on definition classes that
+    # may not be shared.  In other words, those fields are never used directly,
+    # and are always recreated as toplevel fields.  On those fields, the base
+    # setup is useless, because only field.args is used for setting up other
+    # fields.  We therefore skip the base setup for those fields.  The only
+    # attributes of those fields are: '_sequence', 'args', 'model_name', 'name'
+    # and '_module', which makes their __dict__'s size minimal.
+
+    def __set_name__(self, owner, name):
```
Comment (Xavier-Do) [inline odoo/fields.py:306]: It's doing more than setting name

File: odoo/models.py [hunk 5]
```diff
@@ -2856,6 +2874,9 @@ def _setup_base(self):
 
         cls._setup_done = True
 
+        for field in cls._fields.values():
+            field.prepare_setup()
```
Comment (Xavier-Do) [inline odoo/models.py:2878]: Is the method prepare_setup only here to avoid accessing the _private_ attribute `_setup_done`?

**Review by Xavier-Do — COMMENTED** [review]
Small remarks after first lecture.

File: odoo/fields.py (hunk 3 above)
Comment (rco-odoo) [reply odoo/fields.py:347]: I renamed this parameter *on purpose*, since its type is different from the former one:
* before: `model` is an instance of the registry class of the given model
* after: `owner` is a model class (definition class or registry class) that holds the given field

> Also, is this really useful since model_name and name are set in __set_name__.
In theory, it shouldn't be useful.  In practice, it is.  For instance https://github.com/odoo/odoo/blob/14.0/addons/mail/wizard/mail_blacklist_remove.py#L10

File: odoo/fields.py (hunk 4 above)
Comment (rco-odoo) [reply odoo/fields.py:306]: As the docstring says, it performs the base setup of the field.

Comment (rco-odoo) [reply odoo/fields.py:306]: BTW, this method is a special method of Python, see https://docs.python.org/3/reference/datamodel.html?highlight=__set_name__#object.__set_name__

File: odoo/models.py (hunk 5 above)
Comment (rco-odoo) [reply odoo/models.py:2878]: I wanted to hide what is actually done in `field` for setup.
This API allows me to change the setup implementation, if I want to.

**Review by Julien00859 — COMMENTED** [review]
review in progress...

**Comment by fmdl** [comment]
@rco-odoo can you explain what is the new attribute `recursive` ?

File: odoo/fields.py (hunk 3 above)
Comment (Xavier-Do) [reply odoo/fields.py:347]: Ok then there are just some _get_attrs override that need to be adapted the same way in this case. (reminder)

File: odoo/fields.py [hunk 6]
```diff
@@ -211,6 +211,7 @@ def _read_group_many2one_field(self, records, domain, order):
     _modules = None                     # modules that define this field
     _setup_done = None                  # the field's setup state: None, 'base' or 'full'
     _sequence = None                    # absolute ordering of the field
+    _base_fields = ()                   # the fields defining self, in override order
```
[the remark this answers is not in the document]
Comment (Elkasitu) [reply odoo/fields.py:214]: I think it should be clarified/documented in the comment block in which registry and definition classes are explained or somewhere, a few characters really don't help in understanding what each of these attributes do.

IMO `BaseModel._base_fields` could be renamed to something like `_field_definitions` or `_defined_fields` and for `Field._base_fields` idk `_overrides` or `_definitions`?, whatever the name I think they should be different between `BaseModel` and `Field`

File: odoo/models.py [hunk 7]
```diff
@@ -611,10 +617,10 @@ def _build_model(cls, pool, cr):
                 '_name': name,
                 '_register': False,
                 '_original_module': cls._module,
-                '_inherit_module': dict(),              # map parent to introducing module
+                '_inherit_module': {},                  # map parent to introducing module
                 '_inherit_children': OrderedSet(),      # names of children models
                 '_inherits_children': set(),            # names of children models
-                '_fields': OrderedDict(),               # populated in _setup_base()
+                '_fields': {},                          # populated in _setup_base()
```
[the remark this answers is not in the document]
Comment (Xavier-Do) [reply odoo/models.py:612]: dict is actually faster so it is quite interresting to use it. I'm not aware of any difference between the two, could be interresting to confirm ;)
```python
>>> struct=dict
>>> def test():
...     d = struct()
...     for i in range(1000):
...         d[i] = i
... 
>>> timeit.timeit(test, number=1000)
0.08728151675313711
>>> from collections import OrderedDict
>>> struct = OrderedDict
>>> timeit.timeit(test, number=1000)
0.11891509406268597

```

File: odoo/models.py [hunk 8]
```diff
@@ -588,16 +599,11 @@ def _build_model(cls, pool, cr):
...
-        # determine inherited models
-        parents = cls._inherit
-        parents = [parents] if isinstance(parents, str) else (parents or [])
-
-        # determine the model's name
-        name = cls._name or (len(parents) == 1 and parents[0]) or cls.__name__
-
         # all models except 'base' implicitly inherit from 'base'
+        name = cls._name
+        parents = list(cls._inherit)
```
[the remark this answers is not in the document]
Comment (Elkasitu) [reply odoo/models.py:593]: it is converted to `["foo"]` in `BaseModel.__new__()`

**Comment by Elkasitu** [comment]
> @rco-odoo can you explain what is the new attribute `recursive` ?
A recursive field is a compute field that depends on itself in some capacity in order to perform the computation.

It's not really new in the sense that the attribute existed prior to this patch, the difference was that it was dynamically set (the ORM did the job of figuring out which fields were recursive and which ones weren't) and now it's done statically (when defining the field). The reason why it's done statically now is because doing it dynamically would rely on components that make the sharing of the field impossible (which reduces bootstrapping performance)

**Comment by Xavier-Do** [comment]
> > @rco-odoo can you explain what is the new attribute `recursive` ?
> 
> To the best of my knowledge, a recursive field is a compute field that depends on itself in some capacity in order to perform the computation.
> 
> It's not really new in the sense that the attribute existed prior to this patch, the difference was that it was dynamically set (the ORM did the job of figuring out which fields were recursive and which ones weren't) and now it's done statically (when defining the field). 

:+1: 

> The reason why it's done statically now is because doing it dynamically would rely on components that make the sharing of the field impossible (which reduces bootstrapping performance)

It is still a little sad that it isn't detected automatically since most dev will forget to add this attribute. I may have missed that but is it/would it be possible to add a check on install/update in case someone forget that? I'm pretty sure we have some recursive field on runbot :D

**Comment by Elkasitu** [comment]
> It is still a little sad that it isn't detected automatically since most dev will forget to add this attribute. I may have missed that but is it/would it be possible to add a check on install/update in case someone forget that? I'm pretty sure we have some recursive field on runbot :D

There is a warning, not sure exactly when it's emitted though

**Comment by Xavier-Do** [comment]
> > It is still a little sad that it isn't detected automatically since most dev will forget to add this attribute. I may have missed that but is it/would it be possible to add a check on install/update in case someone forget that? I'm pretty sure we have some recursive field on runbot :D
> 
> There is a warning, not sure exactly when it's emitted though

Great news I'll test that.

File: odoo/models.py [hunk 9]
```diff
@@ -251,6 +251,93 @@ def expand_ids(id0, ids):
...
+# model's instances for a given registry.
+#
+#       class A1(Model):                      Model
+#           _name = 'a'                       / | \
+#                                            A3 A2 A1   <- definition classes
+#       class A2(Model):                      \ | /
+#           _inherit = 'a'                      a       <- registry class
+#                                               |
+#       class A3(Model):                     records
+#           _inherit = 'a'
```
[the remark this answers is not in the document]
Comment (rco-odoo) [reply odoo/models.py:318]: That comment is great, because it shows that you confuse `registry['a']` (the class) with `env['a']` (an instance)...

File: odoo/models.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rco-odoo) [reply odoo/models.py:612]: The difference between `dict` and `OrderedDict` is that the latter can *reorder* the keys: you have methods to move a given key in first and last position, respectively.

File: odoo/models.py [hunk 10]
```diff
@@ -338,7 +347,7 @@ def expand_ids(id0, ids):
 # the given registries.
 
 
-class BaseModel(MetaModel('DummyModel', (object,), {'_register': False})):
+class BaseModel(metaclass=MetaModel):
```
[the remark this answers is not in the document]
Comment (rco-odoo) [reply odoo/models.py:391]: Since Odoo 13.0, we require Python ≥ 3.6

File: odoo/fields.py [hunk 11]
```diff
@@ -274,43 +275,35 @@ def __repr__(self):
...
         """ Return the field parameter attributes as a dictionary. """
         # determine all inherited field attributes
         attrs = {}
         modules = []
         for field in self.args.get('_base_fields', ()):
-            if not self._can_setup_from(field):
+            if not isinstance(field, type(self)):
                 attrs.clear()
                 modules.clear()
                 continue
```
[the remark this answers is not in the document]
Comment (rco-odoo) [reply odoo/fields.py]: The idea is actually to not take into account the parameters of `field` if it is not the same type as `self`.
For instance, this deals with something like below.  The idea is: on the registry class, the field will be an integer, so ignore the parameters from the many2one field definition.
```python
class A(Model):
    _name = 'a'
    foo_id = fields.Many2one('b', required=True)

class B(Model):
    _inherit = 'a'
    foo_id = fields.Integer()
```
I am not even sure we have a concrete use-case in Odoo for this.

File: odoo/models.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/models.py:318]: Ah... those recordsets again...

File: odoo/addons/base/models/ir_config_parameter.py [hunk 12]
```diff
@@ -63,13 +63,16 @@ def get_param(self, key, default=False):
...
     @api.model
-    @ormcache('self.env.uid', 'self.env.su', 'key')
+    @ormcache('key')
     def _get_param(self, key):
-        params = self.search_read([('key', '=', key)], fields=['value'], limit=1)
-        return params[0]['value'] if params else None
+        self.flush(['key', 'value'])
+        self.env.cr.execute("SELECT value FROM ir_config_parameter WHERE key = %s", [key])
+        result = self.env.cr.fetchone()
+        return result and result[0]
```
[the remark this answers is not in the document]
Comment (rco-odoo) [reply odoo/addons/base/models/ir_config_parameter.py:75]: This is a bug fix for https://github.com/odoo-dev/odoo/blob/master-field-setup-rco/addons/crm/models/crm_lead.py#L482
Setting up this field from `crm.lead` requires to use the ORM to retrieve some `ir.config_parameter`, but the registry isn't ready yet, so it crashes.

File: odoo/models.py [hunk 13]
```diff
@@ -2882,11 +2887,17 @@ def _setup_base(self):
...
-            for name, field in sorted(getmembers(cls, Field.__instancecheck__), key=lambda f: f[1]._sequence):
-                # do not retrieve magic, custom and inherited fields
-                if not any(field.args.get(k) for k in ('automatic', 'manual', 'inherited')):
-                    self._add_field(name, field.new())
+            cls._fields.clear()
+
+            # collect the definitions of each field (base definition + overrides)
+            definitions = defaultdict(list)
+            for klass in reversed(cls.mro()):
+                if getattr(klass, 'pool', None) is None and isinstance(klass, MetaModel):
```
[the remark this answers is not in the document]
Comment (rco-odoo) [reply odoo/models.py]: What about setting explicitly `pool = None` on `BaseModel`?
On definition classes `klass.pool` is `None`, while on registry classes `klass.pool` is the class' registry.
For the record, in ancient times, the registry was called the "pool", because it is a... collection of... objects :laughing:

**Comment by rco-odoo** [comment]
@robodoo rebase-merge r+
